Added multipart upload in S3 storage adapter

vershwal opened the pull request:
ref https://linear.app/ghost/issue/PRO-1538

- GCS doesn't support AWS streaming signature format, causing uploads to fail with SignatureDoesNotMatch. This can be fixed by using buffer instead of stream for uploads. But this will cause memory issues for large files
- This commit implements a multipart upload for files > 10MB
- Small files (<10MB): Use buffer upload (fs.readFile)
- Large files (≥10MB): Use S3 multipart upload API for memory efficiency. Multipart uploads read file in 10MB chunks, keeping RAM usage constant
- Added automatic cleanup with AbortMultipartUpload on errors
- Multipart upload is S3-standard and works with both AWS S3 and GCS

vershwal committed:
Fixed GCS uploads with buffer uploads and multipart support

ref https://linear.app/ghost/issue/PRO-1538

- GCS doesn't support AWS streaming signature format, causing uploads to fail with SignatureDoesNotMatch. This can be fixed by using buffer instead of stream for uploads. But this will cause memory issues for large files
- This commit implements a multipart upload for files > 10MB
- Small files (<10MB): Use buffer upload (fs.readFile)
- Large files (≥10MB): Use S3 multipart upload API for memory efficiency. Multipart uploads read file in 10MB chunks, keeping RAM usage constant
- Added automatic cleanup with AbortMultipartUpload on errors
- Multipart upload is S3-standard and works with both AWS S3 and GCS.

vershwal committed:
Updated errors and logs

vershwal committed:
Fixed lint

vershwal committed:
Added tests for multipart upload

vershwal committed:
Merge branch 'main' into gcsMultipartUpload

allouis commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -48,6 +56,8 @@ export interface S3StorageOptions {
     tenantPrefix?: string;
     s3Client?: S3Client;
+    multipartThreshold?: number; // File size threshold for multipart upload (default: 10MB)
+    partSize?: number; // Size of each part for multipart upload (default: 10MB)
```
I think we can remove comments here, we can make the names more explicit if we're worried about losing context - maybe `multipartUploadThresholdBytes` and `multipartChunkSizeBytes` would capture everything?

allouis commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -115,6 +134,17 @@ export default class S3Storage extends StorageBase {
+            return await this.uploadMultipart(file, key, stats.size);
+        }
+ 
+        // Use simple upload for small files (faster, less overhead)
```
Can remove this comment

allouis commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -127,6 +157,106 @@ export default class S3Storage extends StorageBase {
+        let uploadId: string | undefined;
+
+        try {
+            // Step 1: Initiate multipart upload
```
Can remove this comment

allouis commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -127,6 +157,106 @@ export default class S3Storage extends StorageBase {
+                    partNumber += 1;
+
+                    const progress = Math.round((uploadedBytes / fileSize) * 100);
+                    logging.info(`Uploaded part ${partNumber - 1}/${Math.ceil(fileSize / this.partSize)} (${progress}%)`);
```
I think moving this before the increment makes it easier to understand, I couldn't immediately see why we do `+ 1`

allouis commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -127,6 +157,106 @@ export default class S3Storage extends StorageBase {
+                        Key: key,
+                        UploadId: uploadId,
+                        PartNumber: partNumber,
+                        Body: buffer.slice(0, bytesRead)
```
Or even we can pull that looping out into a separate method:
```ts
async function* readFileInParts(filePath: string, partSize: number): AsyncGenerator<Buffer> {
    const stream = fs.createReadStream(filePath);
    let buffer = Buffer.alloc(0);

    for await (const chunk of stream) {
         buffer = Buffer.concat([buffer, chunk as Buffer]);

         while (buffer.length >= partSize) {
              yield buffer.slice(0, partSize);
              buffer = buffer.slice(partSize);
         }
    }

    if (buffer.length > 0) {
         yield buffer;
    }
}
```

And then we can just do something like

```ts
const parts = this.readFileInParts(filePath, partSize);

for await (const part of parts) {
    // part is a buffer of the correct size and we can upload it
}
```

vershwal replied:
Good call! I've refactored this to use a stream-based approach with an async generator. The upload loop is now much cleaner. This is definitely easier to understand than the manual fileHandle.read() with offset tracking

vershwal committed:
Removed redundant comments

vershwal committed:
Minor cleanups

- Made multipart options required
- Improved logging

vershwal committed:
Refactored S3 multipart upload to use stream-based file reading

- Replaced manual file handle reading with async generator using fs.createReadStream
- Added readFileInChunks() method that yields correctly-sized buffers
- Removed manual buffer slicing and position tracking
- Updated Tests

vershwal commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -127,6 +157,106 @@ export default class S3Storage extends StorageBase {
+                throw new errors.InternalServerError({
+                    message: tpl(messages.multipartUploadInitFailed)
+                });
+            }
```
ya, make sense

vershwal committed:
Added validation for AWS S3 minimum multipart upload part size

vershwal committed:
Merge branch 'main' into gcsMultipartUpload

allouis reviewed (commented):
Few small comments but this is looking good!

vershwal commented on ghost/core/core/server/adapters/storage/S3Storage.ts:
```diff
@@ -18,6 +18,9 @@ import {
     S3ClientConfig
 } from '@aws-sdk/client-s3';
 
+// Minimum chunk size for multipart uploads (5 MiB) - required by S3/GCS
```
Yes, GCS has the limit too: https://docs.cloud.google.com/storage/quotas#requests

vershwal committed:
Clean up

- Added gcs limits doc
- Removed redundant logging
- Removed extra if check
